internal/models/one_time_token.go

```diff
@@ -0,0 +1,330 @@
+	return nil
+}
+
+type OneTimeTokenNotFoundError struct {
+}
+
+func (e OneTimeTokenNotFoundError) Error() string {
+	return "One-time token not found"
+}
+
+type OneTimeToken struct {
+	ID uuid.UUID `json:"id" db:"id"`
+
+	UserID    uuid.UUID        `json:"user_id" db:"user_id"`
+	TokenType OneTimeTokenType `json:"token_type" db:"token_type"`
+
+	TokenHash string `json:"token_hash" db:"token_hash"`
+	RelatesTo string `json:"relates_to" db:"relates_to"`
+
+	CreatedAt time.Time `json:"created_at" db:"created_at"`
+	UpdatedAt time.Time `json:"updated_at" db:"updated_at"`
+}
+
+func (OneTimeToken) TableName() string {
+	return "one_time_tokens"
+}
+
+func ClearAllOneTimeTokensForUser(tx *storage.Connection, userID uuid.UUID) error {
+	return tx.Q().Where("user_id = ?", userID).Delete(OneTimeToken{})
+}
+
+func ClearOneTimeTokenForUser(tx *storage.Connection, userID uuid.UUID, tokenType OneTimeTokenType) error {
+	if err := tx.Q().Where("token_type = ? and user_id = ?", tokenType, userID).Delete(OneTimeToken{}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func CreateOneTimeToken(tx *storage.Connection, userID uuid.UUID, relatesTo, tokenHash string, tokenType OneTimeTokenType) error {
+	if err := ClearOneTimeTokenForUser(tx, userID, tokenType); err != nil {
+		return err
+	}
+
+	oneTimeToken := &OneTimeToken{
+		ID:        uuid.Must(uuid.NewV4()),
+		UserID:    userID,
+		TokenType: tokenType,
+		TokenHash: tokenHash,
+		RelatesTo: strings.ToLower(relatesTo),
+	}
+
+	if err := tx.Eager().Create(oneTimeToken); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func FindOneTimeToken(tx *storage.Connection, tokenHash string, tokenTypes ...OneTimeTokenType) (*OneTimeToken, error) {
+	oneTimeToken := &OneTimeToken{}
+
+	query := tx.Eager().Q()
+
+	switch len(tokenTypes) {
+	case 2:
+		query = query.Where("(token_type = ? or token_type = ?) and token_hash = ?", tokenTypes[0], tokenTypes[1], tokenHash)
+
+	case 1:
+		query = query.Where("token_type = ? and token_hash = ?", tokenTypes[0], tokenHash)
+
+	default:
+		panic("FindOneTimeToken accepts only 3 or 4 arguments")
+	}
+
+	if err := query.First(oneTimeToken); err != nil {
+		if errors.Cause(err) == sql.ErrNoRows {
+			return nil, OneTimeTokenNotFoundError{}
+		}
+
+		return nil, errors.Wrap(err, "error finding one time token")
+	}
+
+	return oneTimeToken, nil
+}
+
+// FindUserByConfirmationToken finds users with the matching confirmation token.
+func FindUserByConfirmationOrRecoveryToken(tx *storage.Connection, token string) (*User, error) {
+	ott, err := FindOneTimeToken(tx, token, ConfirmationToken, RecoveryToken)
+	if err != nil && !IsNotFoundError(err) {
+		return nil, err
+	}
+
+	if ott == nil {
+		user, err := findUser(tx, "(confirmation_token = ? or recovery_token = ?) and is_sso_user = false", token, token)
+		if err != nil {
+			if IsNotFoundError(err) {
+				return nil, ConfirmationOrRecoveryTokenNotFoundError{}
+			} else {
+				return nil, err
+			}
+		}
+
+		return user, nil
+	}
```

Comment:
Phase II means removing this.

Reply:
can you add a TODO comment so that it's clear what needs to be removed next? worried that this will get lost in the PR review

Reply:
Already opened a PR, so no need for a todo.